### DISCUSSION TITLE
Parse tokens based on previously parsed tokens

### DISCUSSION THREAD

**Discussion by FoamScience** [tree-sitter/tree-sitter#1477] [Q&A] [answered]
Hi guys,

So, I want to build a grammar to parse something like this ("number" and "vector" are actually mentioned exactly like this in the parsed file):
```
List<number> 2 ( 3  4 );
List<vector>   2 ( (1 2 3) (4 5 6) );
List<word> 2 ( some thing );
```

I can parse numbers and vectors just fine and I was wondering if tree-sitter supports a way to generically
parse these list (instead of writing numberList and vectorList parsers ... etc)

Thanks in advance for your help!

**Comment by alemuller** [comment]
I'm not sure if I quite understood your problem.


Do you mean something like this? If not, could you provide an example of the tree that you are tying to generate?
```js
    rules: {

        grammar: $ => repeat($.foo),

        foo: $ => seq(
          $.aggregate, ';'
        ),

        aggregate: $ => seq(
          '(',
          repeat(choice(
            $.number,
            $.aggregate
          )),
          ')'
        ),

      number: $ => /[0-9]+/,
    }
```
```
( 3 4 );
( (1 2 3) (4 5 6) );
```
```
(grammar [0, 0] - [2, 0]
  (foo [0, 0] - [0, 8]
    (aggregate [0, 0] - [0, 7]
      (number [0, 2] - [0, 3])
      (number [0, 4] - [0, 5])))
  (foo [1, 0] - [1, 20]
    (aggregate [1, 0] - [1, 19]
      (aggregate [1, 2] - [1, 9]
        (number [1, 3] - [1, 4])
        (number [1, 5] - [1, 6])
        (number [1, 7] - [1, 8]))
      (aggregate [1, 10] - [1, 17]
        (number [1, 11] - [1, 12])
        (number [1, 13] - [1, 14])
        (number [1, 15] - [1, 16])))))
```

**Reply by FoamScience** [reply]
While `aggregate` is going in the right direction, the problem is list elements have different types which I want to parse too if possible;
This is similar to parsing `initializer_list` in C. I added a list of words to make things clearer;
A sample tree can be of the form (It recognizes if a list element is a vector, and applies the parser for "vector" there) :
```
(grammar [0, 0] - [2, 0]
  (foo [0, 0] - [0, 8]
    (aggregate [0, 0] - [0, 7]
      (number [0, 2] - [0, 3])
      (number [0, 4] - [0, 5])))
  (foo [1, 0] - [1, 20]
    (aggregate [1, 0] - [1, 19]
      (vector[1, 2] - [1, 9]
        (number [1, 3] - [1, 4])
        (number [1, 5] - [1, 6])
        (number [1, 7] - [1, 8]))
      (vector [1, 10] - [1, 17]
        (number [1, 11] - [1, 12])
        (number [1, 13] - [1, 14])
        (number [1, 15] - [1, 16])))))
```

**Comment by alemuller** [comment, marked as answer]
Nested parenthesis are common in lisp language, such [emacs lisp](https://github.com/Wilfred/tree-sitter-elisp),[fennel](https://github.com/travonted/tree-sitter-fennel), [s-exp ](https://github.com/AbstractMachinesLab/tree-sitter-sexp), and others. Probably should see how they implement this.

```js
    grammar: $ => repeat($.foo),

    foo: $ => seq(
        $.list, ';'
    ),

    list: $ => seq(
        '(',
        repeat1($._listExpr),
        ')'
    ),

    _listExpr: $ => choice(
        $._expr,
        $.vector,
        // something else unique to list:
        //$.somethingElseList
    ),

    vector: $ => seq(
        '(',
        repeat1($._vectorExpr),
        ')'
    ),

    _vectorExpr: $ => choice(
        $._expr,
        // something else unique to vector:
        //$.somethingElseVec
    ),

    _expr: $ => choice(
        $.number,
    ),

    number: $ => /[0-9]+/,
```

**Reply by FoamScience** [reply]
Thanks, so I sadly have to `choice($.number, $.vector ...)` which I was trying to avoid because of laziness :)

**Reply by alemuller** [reply]
You can `alias($.list, $.vector)` on `_listExpr` if the both accepts the same expression.
